proposals/records/1-index-scaling.md

```diff
@@ -0,0 +1,183 @@
+- Index Consumer: Index data against all shard replicas and validate consistency
+- Search proxy: Search requests go to this component, which knows about all
+  IndexNodes and which one is best to perform search against
+- Increase/descrease replica count by individual shard and move shards around depending on IndexNode load
+- Manage shard reindexing / shadow shard implementation 
+
+This proposal focuses on an implementation that will cause minimal impact on the existing
+architecture and allows us to test/gather feedback on feasability quickly.
+
+### IndexNode Shard replication:
+
+The IndexNodeCoordinator will ask 1 IndexNode to copy it's shard data to another IndexNode:
+
+
+```mermaid
+flowchart TB
+    IndexNodeCoordinator-- CopyShardRequest -->SourceIndexNode
+    SourceIndexNode-- CopyShard -->DestinationIndexNode
+    IndexNodeCoordinator-- ValidateShard -->DestinationIndexNode
+```
+
+When an IndexNode needs it's replicas increased:
+
+- All indexing to shard needs to be buffered
+- Reads can only go to existing known good shards
+
+### Indexing:
+
+A single NATS stream will be used for indexing data instead
+of the ingest component needing to know about and produce
+messages to many streams depending on replica count.
+
+The IndexNodeCoordinator will be responsible for fanning
+out writes to replicas.
+
+When fully implemented, ingest will not need to know about
+how many replicas a shard has.
+
+```mermaid
+flowchart TB
+    Ingest-- Index data -->NATS
+    NATS-- consume index data --->IndexNodeCoordinator
+    IndexNodeCoordinator-- Index --->IndexNodeReplica1
+    IndexNodeCoordinator-- Index --->IndexNodeReplica2
+    IndexNodeCoordinator-- Index --->IndexNodeReplica3
+```
+
+### Searches
+
+Search requests will also now go through the IndexNodeCoordinator instead of the `search` component
+needing to know about each `IndexNode` and number of replicas.
+
+```mermaid
+flowchart TD
+    Search --> IndexNodeCoordinator
+    IndexNodeCoordinator --> KBIndexNodeReplica1
+    IndexNodeCoordinator --> KBIndexNodeReplica2
+    IndexNodeCoordinator --> KBIndexNodeReplica3
+```
+
+### IndexNode metrics
+
+An important part of this proposal is to be tracking IndexNode
+and replica statistics.
+
+These statistics will be used for determining:
+- How shards should be distributed across IndexNodes
+- Where new replicas should be placed
+- How to rebalance shards
+- When to allocate new shards to scale searches
+
+
+## Rollout plan
+
+### Phase 1: Implement side-by-side
+
+Phase 1 will focus on an approach with limited impact on the existing implementation.
+
+- Existing IndexNodes and architecture will stay in place
+- New IndexCoordinator component provided
+- New IndexNodes deployed seperately from existing nodes
+- Validate indexing, search approach, feature flag for test accounts
+
+### Phase 2: Cutover
+
+Once phase 1 is validated, cut over new scheme.
```

Comment:
How do you imagine the cut over? Would it be kind of a migration script that assings old IndexNode data to IndexCoordinators?

Reply:
In phase 1, we should index all the shards to the new coordinator nodes so this will just be a cutover. Everything should  already be validated as working.